bin/PEPS.py

```diff
@@ -0,0 +1,109 @@
+import torch
+import numpy as np
+import opt_einsum as oe
+
+
+class PEPS():
+    def __init__(self, physical_labels, virtual_lables, vertex_labels,
+                 vertex_tensors):
+        self.physical_labels = physical_labels
+        self.virtual_lables = virtual_lables
+        self.vertex_labels = vertex_labels
+        self.vertex_tensors = vertex_tensors
+        self.N = len(self.vertex_tensors)
+        self.normalize()
+
+    def get_einsum_order(self):
+        """ There is a problem oe.contract_expression, need to check. Workaround with on the fly contraction
+        order for now"""
+        # Statetensor
+        sublist = []
+        for i,t in enumerate(self.vertex_tensors):
+            sublist.append(self.vertex_labels[i])
+            sublist.append(t.shape)
+        print(sublist)
+        self.code_statetensor = oe.contract_expression(*sublist)
+        # Inner product
+        sublist = []
+        largest_label = self.virtual_lables.max()
+        for i,t in enumerate(self.vertex_tensors):
+            sublist.append(self.vertex_labels[i])
+            sublist.append(t.shape)            
+            t_c_labels = self.vertex_labels[i] + largest_label
+            t_c_labels[-1] = self.vertex_labels[i][-1]
+            sublist.append(t_c_labels)
+            sublist.append(t.shape)
+        self.code_inner_product = oe.contract_expression(*sublist)
+
+    def copy(self):
+        return PEPS(self.physical_labels,
+                    self.virtual_lables,
+                    self.vertex_labels,
+                    [t.clone() for t in self.vertex_tensors])
+
+    def conj(self):
+        for i in range(self.N):
+            self.vertex_tensors[i] = self.vertex_tensors[i].conj()
+        return self
+
+    def output_variables(self):
+        return torch.cat([t.flatten() for t in self.vertex_tensors])
+
+    def load_variables(self, variables):
```

Comment:
pytorch has its own parameter management system, you'd better use its own. It is available in `Module` system, for an example,
https://github.com/TensorBFS/dTRG/blob/312ff05cd40b3a3f7b0237b878d6f3275dd054d1/dTRG.py#L65